Support Google Pub/Sub source

saiharshavellanki opened the pull request:
Added support for Google Pub/Sub data source
Updated the documentation

<img width="1461" alt="Screenshot 2024-04-24 at 12 07 25 PM" src="https://github.com/fennel-ai/client/assets/20837258/37ac6c51-79bc-4ddc-8dbe-6bac6bc7db9c">

<img width="1447" alt="Screenshot 2024-04-24 at 12 07 38 PM" src="https://github.com/fennel-ai/client/assets/20837258/ebb44f29-fb7a-424c-81a6-fab89d53836a">

<img width="1437" alt="Screenshot 2024-04-24 at 12 07 50 PM" src="https://github.com/fennel-ai/client/assets/20837258/5cd554cf-9cde-438e-89f4-196b679aaf49">

nikhilgarg28 commented on docs/examples/api-reference/sources/sql.py:
```diff
@@ -218,3 +218,40 @@ class UserClick:
+    # docsnip-highlight end
+
+    # docsnip-highlight next-line
+    collection = pubsub.topic("test_topic")
```
Rename `collection` to `topic`. In fact, try to inline `pubsub.topic(...)` in the `@source` decorator in a single line - if you can, do that. Else, keep it like this across two lines. You may have to shorten `test_topic` to say `name`

nikhilgarg28 commented on docs/pages/api-reference/connectors/pubsub.md:
```diff
@@ -0,0 +1,70 @@
+</Expandable>
+
+<Expandable title="service_account_key" type="str">
+The JSON string containing the credentials for the Service Account to use to access
+Pub/Sub topic. See below for instructions on how to obtain this.
```
This isn't a json string but a dictionary, right? If so, please change this line to mention that.

saiharshavellanki replied:
I copied this from BigQuery and forgot to update. Updated now

nikhilgarg28 commented on docs/pages/api-reference/connectors/pubsub.md:
```diff
@@ -0,0 +1,70 @@
+Pub/Sub topic. See below for instructions on how to obtain this.
+</Expandable>
+
+:::info
+Fennel supports only Upsert mode CDC with data in JSON format. If you require support
+for schema or CDC data format, please reach out to Fennel support.
+:::
```
I don't follow this. What are we trying to say? Can use not send debezium data in pub-sub? Similarly, can use not do `append` mode only for keyless datasets?

mohitreddy1996 replied:
@nikhilgarg28 there are some limitations with Pubsub inherently around ordered delivery of messages which prevents us from performing exactly-once semantics. We have listed down some of it here - https://coda.io/d/_dmX2EX1n2_s/Pub-Sub-Fennel-Documentation_suRHk

saiharshavellanki replied:
That is correct. Additionally, I am rewording to "supports only Append and Upsert mode with CDC" since we can support Append as well similar to other

nikhilgarg28 commented on docs/pages/api-reference/connectors/pubsub.md:
```diff
@@ -0,0 +1,70 @@
+:::
+
+#### Topic Parameters
+<Expandable title="topic_id" type="str">
```
Is `topic_id` a standard GCP thing? If not, let's just call it `name` since `id` usually denotes an integer, not a string.

mohitreddy1996 replied:
yeah its a pub-sub specific thing - 
![Uploading image.png…]()

saiharshavellanki replied:
Yes that is the standard naming convention

nikhilgarg28 commented on docs/pages/api-reference/connectors/pubsub.md:
```diff
@@ -0,0 +1,70 @@
+
+At this point, you should have a service account with the "Pub/Sub Subscriber" project-level permission.
+
+For Service Account Key JSON, enter the Google
+Cloud [Service Account Key in JSON format](https://cloud.google.com/iam/docs/creating-managing-service-account-keys).
```
This part isn't clear and needs to be reworded. I suppose you mean to simply say something like "To obtain a Service Account Key, follow the instructions on <link to page>" or something like that?

saiharshavellanki replied:
That is correct. Let me reword it

mohitreddy1996 reviewed (changes requested):
@saiharshavellanki I think we should anyways add format as an input to the pub sub topic decorator. We can add validations for now that JSON is the only valid format for now, but let's setup the base for potential highly likely extensions.

saiharshavellanki commented:
> @saiharshavellanki I think we should anyways add format as an input to the pub sub topic decorator. We can add validations for now that JSON is the only valid format for now, but let's setup the base for potential highly likely extensions.

Added format as input to decorator and added validation
